src/kde/quick_index.i.hh

```diff
@@ -0,0 +1,328 @@
+        }
+      }
+    }
+
+    // collect the local ghost locations
+    local_ghost_locations = collect_ghost_data(locations);
+
+  } // End domain decomposed data construction
+}
+
+template <size_t dim>
+std::vector<std::array<double, 3>>
+quick_index<dim>::collect_ghost_data(const std::vector<std::array<double, 3>> &local_data) const {
+  Require(local_data.size() == n_locations);
+  Insist(domain_decomposed, "Calling collect_ghost_data with a quick_index object that specified "
+                            "domain_decomposed=.false.");
+  // allocate the ghost data
+  std::vector<std::array<double, 3>> local_ghost_data(local_ghost_buffer_size, {0.0, 0.0, 0.0});
+  // Use one sided MPI Put commands to fill up the ghost cell location data
+  std::vector<double> local_ghost_buffer(local_ghost_buffer_size);
+  MPI_Win win;
+  MPI_Win_create(&local_ghost_buffer[0], local_ghost_buffer_size * sizeof(double), sizeof(double),
+                 MPI_INFO_NULL, MPI_COMM_WORLD, &win);
+
+  // working from my local data put the ghost data on the other ranks
+  for (size_t d = 0; d < dim; d++) {
+    MPI_Win_fence(0, win);
+    for (auto putItr = put_window_map.begin(); putItr != put_window_map.end(); putItr++) {
+      // use map.at() to allow const access
+      const auto &index_vector = coarse_index_map.at(putItr->first);
+      // fill up the current ghost cell data for this dimension
+      std::vector<double> put_buffer(index_vector.size());
+      int putIndex = 0;
+      for (auto indexItr = index_vector.begin(); indexItr != index_vector.end(); indexItr++) {
+        put_buffer[putIndex] = local_data[*indexItr][d];
+        putIndex++;
+      }
+      // loop over all ranks we need to send this buffer too.
+      for (auto rankItr = putItr->second.begin(); rankItr != putItr->second.end(); rankItr++) {
+        const std::array<int, 3> &putv = *rankItr;
+        int put_rank = putv[0];
+        int put_rank_buffer_size = putv[1];
+        int put_offset = putv[2];
+        MPI_Put(&put_buffer[0], static_cast<int>(put_buffer.size()), MPI_DOUBLE, put_rank,
+                put_offset, put_rank_buffer_size, MPI_DOUBLE, win);
+      }
+    }
+    MPI_Win_fence(0, win);
+    // alright move the position buffer to the final correct array positions
+    int posIndex = 0;
+    for (auto posItr = local_ghost_buffer.begin(); posItr != local_ghost_buffer.end(); posItr++) {
+      local_ghost_data[posIndex][d] = *posItr;
+      posIndex++;
+    }
+  }
+  MPI_Win_free(&win);
+  return local_ghost_data;
+}
+
+template <size_t dim>
+std::vector<double>
+quick_index<dim>::collect_ghost_data(const std::vector<double> &local_data) const {
+  Require(local_data.size() == n_locations);
+  Insist(domain_decomposed, "Calling collect_ghost_data with a quick_index object that specified "
+                            "domain_decomposed=.false.");
+  // allocate the ghost data
+  std::vector<double> local_ghost_data(local_ghost_buffer_size, 0.0);
+  MPI_Win win;
```

Comment:
@KineticTheory I'm not 100% sure the best way to wrap these into c4. Can you help me with this?

Reply:
I think we decided to wait and wrap these mpi calls in a different PR.